File: xml_parser.js
// xml_parser.js
const fs = require('fs');
const util = require('util');

// Simple XML parser using regex (for demo) — in a real app, use a library like xml2js.
// But we'll implement a minimal parser to avoid dependencies.
function parseXML(xml) {
    // Remove comments and processing instructions
    xml = xml.replace(/<!--[\s\S]*?-->/g, '');
    xml = xml.replace(/<\?[\s\S]*?\?>/g, '');
    // Use a simple stack-based parser (not fully robust, but works for basic cases)
    let i = 0;
    let root = null;
    let stack = [];
    let current = null;
    while (i < xml.length) {
        if (xml[i] === '<') {
            if (xml[i+1] === '/') {
                // closing tag
                let closeIndex = xml.indexOf('>', i);
                let tagName = xml.substring(i+2, closeIndex).trim();
                if (stack.length > 0) {
                    let parent = stack.pop();
                    // If text accumulated in current, assign it before closing
                    if (current && current !== parent && current.text !== undefined) {
                        // assign text to current
                    }
                    current = parent;
                }
                i = closeIndex + 1;
            } else if (xml[i+1] === '!') {
                // CDATA or comment (already removed)
                let closeIndex = xml.indexOf('>', i);
                i = closeIndex + 1;
            } else {
                // opening tag
                let closeIndex = xml.indexOf('>', i);
                let tagContent = xml.substring(i+1, closeIndex);
                let tagMatch = tagContent.match(/^(\w+)(?:\s+(.*))?/);
                if (tagMatch) {
                    let tagName = tagMatch[1];
                    let attrStr = tagMatch[2] || '';
                    let attrs = {};
                    // parse attributes (simple regex)
                    let attrRegex = /(\w+)\s*=\s*"([^"]*)"/g;
                    let match;
                    while ((match = attrRegex.exec(attrStr)) !== null) {
                        attrs[match[1]] = match[2];
                    }
                    let node = { tag: tagName, attrs, children: [], text: '' };
                    if (tagContent.endsWith('/')) {
                        // self-closing
                        // add to parent
                        if (stack.length > 0) {
                            stack[stack.length-1].children.push(node);
                        } else {
                            root = node;
                        }
                    } else {
                        if (stack.length > 0) {
                            stack[stack.length-1].children.push(node);
                        } else {
                            root = node;
                        }
                        stack.push(node);
                        current = node;
                    }
                }
                i = closeIndex + 1;
            }
        } else {
            // text node
            let nextOpen = xml.indexOf('<', i);
            let text = xml.substring(i, nextOpen).trim();
            if (text && current) {
                // If current has no children and no text yet, assign
                if (current.children.length === 0 && current.text === '') {
                    current.text = text;
                } else {
                    // If there are children, text might be whitespace; we ignore or add as text node?
                    // For simplicity, if children exist, we ignore whitespace text.
                }
            }
            i = nextOpen === -1 ? xml.length : nextOpen;
        }
    }
    return root;
}

function displayNode(node, indent = 0) {
    const prefix = '  '.repeat(indent);
    let attrStr = Object.entries(node.attrs).map(([k,v]) => `${k}="${v}"`).join(' ');
    attrStr = attrStr ? ' ' + attrStr : '';
    if (node.children.length === 0) {
        if (node.text) {
            console.log(`${prefix}<${node.tag}${attrStr}>${node.text}</${node.tag}>`);
        } else {
            console.log(`${prefix}<${node.tag}${attrStr}/>`);
        }
    } else {
        console.log(`${prefix}<${node.tag}${attrStr}>`);
        node.children.forEach(child => displayNode(child, indent+1));
        console.log(`${prefix}</${node.tag}>`);
    }
}

function toJSON(node) {
    let obj = {};
    if (node.text) obj['#text'] = node.text;
    Object.keys(node.attrs).forEach(k => obj['@'+k] = node.attrs[k]);
    if (node.children.length > 0) {
        let childMap = {};
        node.children.forEach(child => {
            let tag = child.tag;
            let childObj = toJSON(child);
            if (childMap[tag]) {
                if (!Array.isArray(childMap[tag])) {
                    childMap[tag] = [childMap[tag]];
                }
                childMap[tag].push(childObj);
            } else {
                childMap[tag] = childObj;
            }
        });
        Object.keys(childMap).forEach(tag => {
            obj[tag] = childMap[tag];
        });
    }
    return obj;
}

function main() {
    const args = process.argv.slice(2);
    let inputFile = null;
    let tag = null;
    let attr = null;
    let toJSONFlag = false;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-i' || args[i] === '--input') {
            inputFile = args[++i];
        } else if (args[i] === '-t' || args[i] === '--tag') {
            tag = args[++i];
        } else if (args[i] === '-a' || args[i] === '--attr') {
            attr = args[++i];
        } else if (args[i] === '--json') {
            toJSONFlag = true;
        } else if (args[i] === '-h' || args[i] === '--help') {
            console.log('Usage: node xml_parser.js [-i input.xml] [-t tag] [-a key=value] [--json]');
            process.exit(0);
        }
    }

    let xmlContent;
    if (inputFile) {
        try {
            xmlContent = fs.readFileSync(inputFile, 'utf8');
        } catch (e) {
            console.error('Error reading file:', e.message);
            process.exit(1);
        }
    } else {
        xmlContent = fs.readFileSync(0, 'utf8');
    }

    let root;
    try {
        root = parseXML(xmlContent);
        if (!root) throw new Error('Parsing failed');
    } catch (e) {
        console.error('XML Parse Error:', e.message);
        process.exit(1);
    }

    if (tag) {
        let found = false;
        const searchTag = (node) => {
            if (node.tag === tag) {
                displayNode(node);
                found = true;
            }
            node.children.forEach(searchTag);
        };
        searchTag(root);
        if (!found) console.log(`No elements with tag '${tag}' found.`);
    } else if (attr) {
        const [key, value] = attr.split('=');
        if (!key || !value) {
            console.error('Invalid attribute format. Use key=value');
            process.exit(1);
        }
        let found = false;
        const searchAttr = (node) => {
            if (node.attrs[key] === value) {
                displayNode(node);
                found = true;
            }
            node.children.forEach(searchAttr);
        };
        searchAttr(root);
        if (!found) console.log(`No elements with attr '${key}=${value}' found.`);
    } else if (toJSONFlag) {
        console.log(JSON.stringify(toJSON(root), null, 2));
    } else {
        displayNode(root);
    }
}

main();

File: XmlParser.cs
// XmlParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml;
using System.Xml.Linq;
using Newtonsoft.Json;

class XmlParser
{
    static void Main(string[] args)
    {
        string inputFile = null;
        string tag = null;
        string attr = null;
        bool toJson = false;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-i": case "--input": inputFile = args[++i]; break;
                case "-t": case "--tag": tag = args[++i]; break;
                case "-a": case "--attr": attr = args[++i]; break;
                case "--json": toJson = true; break;
                case "-h": case "--help": ShowHelp(); return;
            }
        }

        string xmlContent;
        if (inputFile != null)
        {
            xmlContent = File.ReadAllText(inputFile);
        }
        else
        {
            using (var reader = new StreamReader(Console.OpenStandardInput()))
            {
                xmlContent = reader.ReadToEnd();
            }
        }

        XDocument doc;
        try
        {
            doc = XDocument.Parse(xmlContent);
        }
        catch (XmlException e)
        {
            Console.Error.WriteLine($"XML Parse Error: {e.Message}");
            return;
        }

        if (tag != null)
        {
            var elements = doc.Descendants().Where(e => e.Name.LocalName == tag).ToList();
            if (elements.Count == 0)
                Console.WriteLine($"No elements with tag '{tag}' found.");
            else
                elements.ForEach(e => DisplayElement(e, 0));
        }
        else if (attr != null)
        {
            var parts = attr.Split('=');
            if (parts.Length != 2)
            {
                Console.Error.WriteLine("Invalid attribute format. Use key=value");
                return;
            }
            var key = parts[0];
            var value = parts[1];
            var elements = doc.Descendants().Where(e => e.Attribute(key)?.Value == value).ToList();
            if (elements.Count == 0)
                Console.WriteLine($"No elements with attr '{key}={value}' found.");
            else
                elements.ForEach(e => DisplayElement(e, 0));
        }
        else if (toJson)
        {
            var json = JsonConvert.SerializeXNode(doc, Formatting.Indented);
            Console.WriteLine(json);
        }
        else
        {
            DisplayElement(doc.Root, 0);
        }
    }

    static void DisplayElement(XElement el, int indent)
    {
        string prefix = new string(' ', indent * 2);
        string attrs = string.Join(" ", el.Attributes().Select(a => $"{a.Name}=\"{a.Value}\""));
        if (!el.HasElements && string.IsNullOrEmpty(el.Value.Trim()))
        {
            Console.WriteLine($"{prefix}<{el.Name}{(!string.IsNullOrEmpty(attrs) ? " " + attrs : "")}/>");
        }
        else if (!el.HasElements)
        {
            Console.WriteLine($"{prefix}<{el.Name}{(!string.IsNullOrEmpty(attrs) ? " " + attrs : "")}>{el.Value.Trim()}</{el.Name}>");
        }
        else
        {
            Console.WriteLine($"{prefix}<{el.Name}{(!string.IsNullOrEmpty(attrs) ? " " + attrs : "")}>");
            foreach (var child in el.Elements())
                DisplayElement(child, indent + 1);
            Console.WriteLine($"{prefix}</{el.Name}>");
        }
    }

    static void ShowHelp()
    {
        Console.WriteLine("Usage: dotnet run -- -i input.xml [-t tag] [-a key=value] [--json]");
    }
}
